synapse/storage/databases/main/event_federation.py

```diff
@@ -1252,17 +1252,39 @@ async def prune_staged_events_in_room(
         referenced_events: Set[str] = set()
         seen_events: Set[str] = set()
         for row in rows:
-            seen_events.add(row["event_id"])
+            event_id = row["event_id"]
+            seen_events.add(event_id)
             event_d = db_to_json(row["event_json"])
 
             # We don't bother parsing the dicts into full blown event objects,
             # as that is needlessly expensive.
+
+            # We haven't checked that the `prev_events` have the right format
+            # yet, so we check as we go.
+            prev_events = event_d.get("prev_events", [])
+            if not isinstance(prev_events, list):
+                logger.info("Invalid prev_events for %s", event_id)
+                continue
```

Comment:
Yeeees, ideally we'd have checked long before here that the `Event` was valid (given we'd already have parsed the JSON into an event object before persistence). I'm slightly hesitant about dropping the "invalid" events here in case we update the format of events and forget to change it here, which would cause us to *sometimes* drop those events (as this function does not get called for each inbound event).